File: src/olcw78 (die yoong)/jsAlgorithmStudy/week09/38.js
/**
 *
 * @param {string[][]} graph
 * @param {string} start
 * @returns
 */
function solution(graph, start) {
  const visited = new Set();
  const adjList = {};
  // 인접 리스트 먼저 만들기
  for (const [from, to] of graph) {
    if (!adjList[from]) adjList[from] = [];
    adjList[from].push(to);
  }

  function dfs(node, visited, res) {
    // 1. 방문 처리
    visited.add(node);
    // 2. 결과 추가
    result.push(node);
    // 3. 연결된 노드가 없으면 탈출
    if (!adjList[node]) return;

    // 4. 연결된 노드가 있음 -> 연결된 노드들을 순회하여 더 깊이 방문
    for (const adj of adjList[node]) {
      // 이미 방문 되었으면 안됨.
      if (!visited.has(adj)) {
        // 더 깊이 방문
        dfs(adj, visited, res);
      }
    }
  }

  const result = [];
  dfs(start, visited, result);
  return result;
}

const r1 = solution(
  [['A', 'B'], ['B', 'C'], ['C', 'D'], ['D', 'E']],
  'A'
);
console.log(r1); // ['A', 'B', 'C', 'D', 'E']

const r2 = solution(
  [['A', 'B'], ['A', 'C'], ['B', 'D'], ['B', 'E'], ['C', 'F'], ['E', 'F']],
  'A'
);
console.log(r2); //  ['A', 'B', 'D', 'E', 'F', 'C']

File: src/olcw78 (die yoong)/jsAlgorithmStudy/week09/38.go
package week09

import "fmt"

func solution(graph [][]string, start string) []string {
	visited := make(map[string]bool)
	var res []string

	// 인접 리스트 만들기
	adjList := make(map[string][]string)
	for _, strings := range graph {
		from, to := strings[0], strings[1]
		adjList[from] = append(adjList[from], to)
	}

	dfs(start, visited, &res, adjList)
	return res
}

// map 은 참조타입으로 포인터로 넘길필요 없음
// res (slice) 는 내부 수정 시에 포인터로 넘기는게 필요
func dfs(
	node string,
	visited map[string]bool,
	res *[]string,
	adjList map[string][]string,
) {
	visited[node] = true

	*res = append(*res, node)
	if _, ok := adjList[node]; !ok {
		return
	}

	for _, adj := range adjList[node] {
		if !visited[adj] {
			dfs(adj, visited, res, adjList)
		}
	}
}

func Run38() {
	r1 := solution([][]string{
		{"A", "B"},
		{"B", "C"},
		{"C", "D"},
		{"D", "E"},
	}, "A")
	fmt.Println(r1) // ["A", "B", "C", "D", "E"]

	r2 := solution([][]string{
		{"A", "B"},
		{"A", "C"},
		{"B", "D"},
		{"B", "E"},
		{"C", "F"},
		{"E", "F"},
	}, "A")
	fmt.Println(r2) // ["A", "B", "D", "E", "F", "C"]
}
